thewalrus/fock_gradients.py

```diff
@@ -0,0 +1,270 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Gradients of Gaussian gates in the Fock representation
+======================================================
+
+**Module name:** :mod:`thewalrus.fock_gradients`
+
+.. currentmodule:: thewalrus.fock_gradients
+
+Contains some the Fock representation of the standard Gaussian gates
+
+
+Fock Gates
+----------
+
+.. autosummary::
+    Xgate
+    Zgate
+    Sgate
+    Rgate
+    S2gate
+    BSgate
+
+Code details
+^^^^^^^^^^^^
+"""
+import numpy as np
+# pylint: disable=unused-import
+from numba import jit, double, complex128
+
+from thewalrus.quantum import fock_tensor
+from thewalrus.symplectic import squeezing, two_mode_squeezing, beam_splitter
+
+
+@jit("double[:,:](double[:,:], double)")
+def grad_Xgate(T, pref):
+    """Calculate the gradient of the Xgate
+    Args:
+        T (array): Array representing the gate
+    Returns:
+        (array): Array representing the gradient of the gate
+    """
+    cutoff, _ = T.shape
+    cutoff -= 1
+    gradT = np.zeros([cutoff, cutoff])
+    for n in range(cutoff):
+        for m in range(cutoff):
+            gradT[n, m] = np.sqrt(m + 1) * T[n, m + 1] * pref
+            if m > 0:
+                gradT[n, m] -= np.sqrt(m) * T[n, m - 1] * pref
+    return gradT
+
+
+def Xgate(x, cutoff, grad=False, hbar=2, r=np.arcsinh(1.0)):
+    r"""
+    Calculates the Fock representation of the Xgate and its gradient
+    Arg:
+        x (float): Parameter of the gate
+        cutoff (int): Fock ladder cutoff
+        grad (boolean): Whether to calculate the gradient or not
+        hbar (float): Value of hbar is the commutation relation
+        r (float): Value of the parameter used internally in fock_tensor.
+    Returns:
+        tuple(G,dG): The Fock representations of the gate and its gradient
+    """
+    nmodes = 1
+    pref = 1.0 / np.sqrt(2 * hbar)
+    alpha = np.array([pref * x])
+    S = np.identity(2 * nmodes)
+
+    if not grad:
+        return fock_tensor(S, alpha, cutoff, r=r).real, None
+
+    T = fock_tensor(S, alpha, cutoff + 1, r=r).real
+    return T[0:cutoff, 0:cutoff], grad_Xgate(T, pref)
```

Comment:
Does the shape of the fock_tensor change depending on whether the gradient is returned or not? Might be a bit confusing to the user. Perhaps the fock_tensor should always be larger than the gradient tensor?

Reply:
the shape does not change. You calculate a slightly bigger tensor so that the gradients are still exact and then return a subtensor of the tenbsor with the expected size.

Reply:
Ah! Got it